File: LeetCode/121. Best Time to Buy and Sell Stock.cpp
// Author : @AdarshGaur

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int mini=prices[0], maxi=0, ans=0;
        for(int i=1; i<prices.size(); i++){
            if(mini>prices[i]){
                ans = max(ans,maxi-mini);
                mini = prices[i];
                maxi = 0;
            }else{
                maxi = max(maxi, prices[i]);
            }
        }
        ans = max(ans, maxi-mini);
        return ans;
    }
};

File: LeetCode/121. Best Time to Buy and Sell Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // we need to maximize the profit for just a single trade only. 
        // in that case we just need to check which could the least price at which 
        // we could've bought the share to sell at (i)th day to make the profit
        // and the maximum profit we could find at any day is our solution
        int minimumBuyPrice = Integer.MAX_VALUE;
        int maxProfit = 0;
        for(int price: prices){
            maxProfit = Math.max(maxProfit, price - minimumBuyPrice);
            minimumBuyPrice = Math.min(minimumBuyPrice, price);
        }
        return maxProfit;
    }
}
